File: DSA/src/py/reverse_nodes_in_k_group.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def _getNode(self, curr, n):
        while curr and n > 0:
            curr = curr.next
            n-=1
        return curr

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        group_prev = dummy
        while True:
            kth_node = self._getNode(group_prev, k)
            if not kth_node:
                break
            group_next = kth_node.next

            prev, curr = group_next, group_prev.next
            while curr != group_next:
                nxt = curr.next
                curr.next = prev
                prev = curr
                curr = nxt

            tmp = group_prev.next
            group_prev.next = kth_node
            group_prev = tmp
        return dummy.next

File: docs/DSA/src/go/reverse_nodes_in_k_group.go
package main

func getKthNode(node *ListNode, k int) *ListNode {
	for node != nil && k > 0 {
		node = node.Next
		k--
	}
	return node
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	dummy := &ListNode{Val: -1, Next: head}
	group_prev := dummy
	var kthNode, group_next, prev, curr, temp *ListNode
	for true {
		kthNode = getKthNode(group_prev, k)
		if kthNode == nil {
			break
		}

		group_next = kthNode.Next
		prev, curr = group_next, group_prev.Next
		for curr != group_next {
			temp = curr.Next
			curr.Next = prev
			prev = curr
			curr = temp
		}

		temp = group_prev.Next
		group_prev.Next = kthNode
		group_prev = temp
	}
	return dummy.Next
}
